ebos/eclbasevanguard.hh

```diff
@@ -379,6 +383,22 @@ public:
     static void setExternalSummaryConfig(Opm::SummaryConfig* summaryConfig)
     { externalEclSummaryConfig_ = summaryConfig; }
 
+
+    /*!
+    * \brief Returns the summary state
+    *
+    * The summary state is a small container object for
+    * computed, ready to use summary values. The values will typically be used by
+    * the UDQ, WTEST and ACTIONX calculations. Observe that all value *have been
+    * converted to the correct output units*.
```

Comment:
I think that the comment about the units is better put to the documentation of the `Opm::SummaryState` class within `opm-common`. out of curiosity: does this mean that the `summaryState_` contains its values in FIELD units for field decks? If so, this has *a lot* of potential for confusion...

Reply:
> does this mean that the summaryState_ contains its values in FIELD units for field decks? 

Yes.

> If so, this has a lot of potential for confusion...

Hopefully you are wrong .... the object in question is a *state* object; i.e. the state of it is updated during the course of the simulation. It therefor needs to have the same lifetime as the simulator, but all *use* of it is confined to opm-common code; so the ugly detail about units should not really leak out.

Reply:
I also hope that I'm wrong. The reason why I fear that I'm not is that the SI-based simulator needs to fill this object, and the output code also must handle these quantities differently than the normal SI-based ones. IMO it would be thus good if the unit conversion was  done internally by this class, i.e. something like `summaryState_.setQuantitySi("FOO", index, value)`.`summaryState_.setQuantityDeckUnits("FOO", index, value)` plus the corresponding getters.

Reply:
This comment is basically a copy from the class documentation. I can remove the stuff about the units Since the usage is internal in opm-common I guess the units are taken care of outside the simulator.

Reply:
> is that the SI-based simulator needs to fill this object, 

No - this is already handled in the output layer, and will continue to be so. This class is not new - we have used for ~one year.

Reply:
> No - this is already handled in the output layer

if the simulator does not need to interact with it what is the point in increasing its entropy by directly exposing it to the simulator instead of keeping it within `EclipseState`, then?

Reply:
> if the simulator does not need to interact with it what is the point in increasing its entropy by directly exposing it to the simulator instead of keeping it within EclipseState, then?

Well - the `EclipseState` object is immutable (or at least close ...), whereas the `SummaryState` object evolves with the simulation. It is *not* something generated from the deck - as the `EclipseState` is.

Reply:
In the long term, the current approach is IMO a recipe for maintenance hell -- or at least for considerable annoyance -- because each and every of the gazillion dynamic parser objects needs to be handled by the simulator and the simulator is not in that business IMHO. Rather, the schedule-dependent objects should be collected by a single class (`DynamicEclipseState`?) which ideally should also provide a pointer to the immutable data like the deck and the eclState that is created from it.

Reply:
> because each and every of the gazillion dynamic parser objects 

Gazillion(?) - such exaggerations makes a meaningful discussion difficult, please refrain from it. We have the following objects implemented and instantiated in the opm-common repository:

**`Deck`**: The first internalization of the input file - should ideally not be used outside of opm-common.

**`EclipseState`**: Everything static assembled from the deck input.

**`Schedule`**: Everything dynamic from the deck input.

**`SummaryConfig`**: An object internalizing all the summary variables requested in the deck.

These objects are instantiated (roughly) as such:
```
Deck <- input_file
EclipseState <- Deck
Schedule <- Deck, EclipseState
SummaryConfig <- Deck, EclipseState, Schedule
```
the `SummaryConfig` class is *a small detail* - and it unfortunate that it is a class that must be managed at top level by the simulator, that is because it requires the `Schedule`object for construction[1]. So currently we have `Gazillion == 4`. 

This PR adds a new[2] class to the simulator. The class in question is implemented, and used, in code which is currently located in opm-common, but it is *not* instantiated there - it does not use any input from the deck, and is not derived from any of the classes in opm-common. Since it needs to have application lifetime, and is a dynamic quantity - it makes sense to manage it from the simulator.

> Rather, the schedule-dependent objects should be collected by a single class 

The only `Schedule` dependent object - apart from `Schedule` itself is the `SummaryConfig`object.

> [....] which ideally should also provide a pointer to the immutable data like the deck and the eclState that is created from it.

I guess you mean "that it is created from"? Yes - that would be possible, but I would consider that an unfortunate coupling in the code.


[1]: Come to think of it - it *could* be internalized in the `Schedule` object - does not really belong there, but that way the object count would go down with one.

[2]: As I have already pointed out the class `SummaryState`is *not* new - it has been used internally in the output code for ~one year.

Reply:
there is also at least the `parseContext` and the `errorGuard` which the simulators need to wrangle with. Anyway, whether you declare a "gazillion" as `> 4` or not is probably a matter of taste: if I'm strict, I'd define "gazillion" as `> 1` in this context.

Anyway, can we please at least agree upon the fact that the number of these objects has only been growing over the last years and that it probably will continue to do so. (That said, I promise to shut up about the current state if you solemnly swear that no new such objects will be introduced after `summaryState` anymore.) 

(note that maybe our main dissonance stems from the fact that you seem to consider the parser as an end to itself, while I see it more as a necessary evil to achieve another goal, i.e., my view of this code is that the file I/O libraries are good if and only if they do not get in the way.)

Reply:
Doesn't introducing concepts like ACTIONX and UDQ already violate the strict decoupling between the simulator and I/O. That said, I don't have any strong opinion on where to store the SummaryState object, as long it doesn't  involve too much extra work.

Reply:
> Anyway, can we please at least agree upon the fact that the number of these objects has only been growing over the last years 

Yes - that is true; along with `flow` becoming more and more capable.

> [...] . (That said, I promise to shut up about the current state if you solemnly swear that no new such objects will be introduced after summaryState anymore.)

Of course I can not swear to anything like that, but these objects are not introduced for fun - or to piss you off. It is done out of necessity - and in the way I think is the best way - the latter can of course be discussed.

Reply:
> that the file I/O libraries are good if and only if they do not get in the way.)

You are of course entitled to that opinion, but without IO there would not be much point in simulating.

Reply:
that's a quite petroleum-centric view, cf. all other ewoms tests/simulators. Also, even for petroleum engineering there would be little point in the I/O code without a simulator. (Seems like we're stuck ;))

Reply:
> [...] but these objects are not introduced for fun - or to piss you off.

I did not imply that. I said that the current design is flawed: these objects should be collected by a class that is tied to the I/O code instead of the simulator. If this class occasionally needs some new data from the simulator, so be it, but currently these objects are kept around at the simulator level for almost no other purpose than to keep the I/O code happy...

Reply:
I actually agree, the core numerical code should not store these objects (EclipseState etc.) at all, they should be created at main() level or equivalent, and passed to the simulator code as necessary. But that design was left behind with the legacy simulator...

Given that we do not at this point refactor this design completely, @totto82's current proposal is fine.

Reply:
> they should be created at main() level or equivalent, and passed to the simulator code as necessary

I fully agree; the point is that currently `eclproblem/vanguard/simulator/...` serve as a de-facto main - that is is the *really broken design* around here.

Reply:
that was not my point, but well. as usual: change the design, I'm happy with anything as long as it only affects flow and nothing else...